### PR TITLE
update to allow only specific keys to be removed

### PR DESCRIPTION
**Pull request by mcraigPW**
It seems like we're losing some pertinent information if _keys are deleted wholesale.  This patch allows the user to indicate those keys which are not needed while not touching other possibly important ones.  Such as `_where` (a full path key) vs. `_requested` or `_requiredBy`

### CODE REVIEW

File: src/removeNPMAbsolutePaths.js [hunk 1]
```diff
@@ -44,7 +44,12 @@ function processFile(filePath, opts) {
       }
 
       Object.keys(obj).forEach((key) => {
-        if (key[0] === '_') {
+        if (!opts.userSpecifiedKeys) {
```
Comment (juanjoDiaz) [inline src/removeNPMAbsolutePaths.js]: This duplicated code unnecessarily.
Also, as mentioned above, `opts.userSpecifiedKeys` is not really necessary. You can just check if `opts.keys` is defined.
Something like:
```
        const shouldBeDeleted = opts.fields ? (opts.fields.indexOf(key) !== -1) : (key[0] === '_');
        if (shouldBeDeleted) {
          delete obj[key];
          writeFile = true;
        }
```

File: test/data/specific_fields/module/package.json [hunk 2]
```diff
@@ -0,0 +1,57 @@
+{
```
Comment (juanjoDiaz) [inline test/data/specific_fields/module/package.json]: Is there any reason to add this file instead of using ` test/data/underscore_fields/module/package.json`??

File: bin/removeNPMAbsolutePaths [hunk 3]
```diff
@@ -14,11 +14,17 @@ const args = process.argv.slice(2);
 const folder = args[0];
 const opts = {
   force: false,
+  userSpecifiedKeys: false,
```
Comment (juanjoDiaz) [inline bin/removeNPMAbsolutePaths]: `userSpecifiedKeys` is not really necessary. You can just check if `keys` is defined (and rename it to `fields` for clarity).

File: bin/removeNPMAbsolutePaths [hunk 4]
```diff
@@ -14,11 +14,17 @@ const args = process.argv.slice(2);
...
+  keys: [],
 };
 
 args.slice(1).forEach((arg) => {
```
Comment (juanjoDiaz) [inline bin/removeNPMAbsolutePaths]: The argument parsing needs some work. This assumes that every single unmatched property is a key to be included.
As in: `removeNPMAbsolutePaths "MY_FOLDER" wrongPropertyThatShouldBeIgnoredButIsNot --force --userSpecific`

This should work so
`removeNPMAbsolutePaths "MY_FOLDER" --force --fields < ... ONLY STUFF HERE IS CONSIDERED AS KEYS ... >` is valid
`removeNPMAbsolutePaths "MY_FOLDER" --fields < ... ONLY STUFF HERE IS CONSIDERED AS KEYS ... > --force ` is valid and `--force` is not considered a property.

So basically stop parsing files as soon as it finds the next option starting with `--`.
This might mean that we can't use the `forEach` method and need a traditional for-loop. Not sure...

File: test/removeNPMAbsolutePaths.js [hunk 5]
```diff
@@ -149,6 +149,28 @@ describe('removeNPMAbsolutePaths.js', function () {
...
+          keys: ['_inBundle', '_where'],
+        };
+        const promise = removeNPMAbsolutePaths(dirPath, opts);
+        return expect(promise).be.fulfilled
+          .then((results) => {
+            expect(results).to.be.an('array').that.have.lengthOf(3);
+            const fileResults = results.find(result => result.filePath === filePath);
+            expect(fileResults).to.include({ success: true, rewritten: true });
+            expect(fileResults.err)
+              .to.not.exist;
```
Comment (juanjoDiaz) [inline test/removeNPMAbsolutePaths.js]: This shouldn't be in it's own line. It should be `expect(fileResults.err).to.not.exist;`

File: test/removeNPMAbsolutePaths.js [hunk 6]
```diff
@@ -149,6 +149,28 @@ describe('removeNPMAbsolutePaths.js', function () {
...
+        const opts = {
+          userSpecifiedKeys: true,
+          keys: ['_inBundle', '_where'],
+        };
+        const promise = removeNPMAbsolutePaths(dirPath, opts);
+        return expect(promise).be.fulfilled
+          .then((results) => {
+            expect(results).to.be.an('array').that.have.lengthOf(3);
+            const fileResults = results.find(result => result.filePath === filePath);
+            expect(fileResults).to.include({ success: true, rewritten: true });
```
Comment (juanjoDiaz) [inline test/removeNPMAbsolutePaths.js:162]: To make this more reliable you probably want to check that `_inBundle` and `_where` are present in the results but, for example, `_from` is not.

**Review by juanjoDiaz — CHANGES_REQUESTED** [review]
Hei,

Thanks for the contribution! This looks really good.

I've added few minor comments.
My only major blocker is the argument parsing function for the CLI.

Also, please update the readme with the new property.

**Comment by juanjoDiaz** [comment]
Also, for my own knowledge, are other underscored fields like `_requested` or `_requiredBy` used in any way? Have you found any issue removing them?
AFAIK, they weren't really necessary for anything.

File: test/data/specific_fields/module/package.json (hunk 2 above)
Comment (mcraigPW) [reply test/data/specific_fields/module/package.json]: There is not.  I was thinking at first that it would modify the file.

File: test/removeNPMAbsolutePaths.js (hunk 5 above)
Comment (mcraigPW) [reply test/removeNPMAbsolutePaths.js]: oops sorry must have just fat-fingered it.

**Comment by mcraigPW** [comment]
> Also, for my own knowledge, are other underscored fields like _requested or _requiredBy used in any way? Have you found any issue removing them?
> AFAIK, they weren't really necessary for anything.

For me there is occasionally helpful information in some fields.  Today, for instance, during one of my upgrades I noticed something had a production dependency on mocha and was able to track it down via the `_requiredBy` field, which led me to realize it's an older module that I need to find a replacement for.  `_requested` might have some useful information about how npm determined what version to install?
The other reason that a minimal deletion is nice for me is that we keep our prod modules in git for now, so removing all the _ fields would be a lot of unnecessary changed being committed.

**Comment by mcraigPW** [comment]
Are there CLI tests?  I realize I didn't think to update those, so I should probably do that if there are some.

File: test/removeNPMAbsolutePaths.js (hunk 6 above)
Comment (mcraigPW) [reply test/removeNPMAbsolutePaths.js:162]: am I missing an example where the tests expect certain content written out?  It looked like from the stubbed `writeFile` that nothing was actually being saved to the file system.

File: bin/removeNPMAbsolutePaths [hunk 7]
```diff
@@ -16,9 +16,23 @@ const opts = {
   force: false,
 };
 
+let inFields = false;
+
 args.slice(1).forEach((arg) => {
```
Comment (juanjoDiaz) [inline bin/removeNPMAbsolutePaths]: I'd rather something here that makes easier to add new properties when needed. Something like:
```
for (let i = 0; i < args.length; i++) {
  const arg = args[i];
  switch(expression) {
    case '--force':
      opts.force = true;
      break;
    case '--fields':
      opts.fields = opts.fields || [];
      while(args[i + 1].slice(0, 2) !== '--') {
        opts.fields.push(args[i++]);
      }
      break;
  }
}
```
So each property is 100% handled in its own `case`

File: test/removeNPMAbsolutePaths.js [hunk 8]
```diff
@@ -406,5 +403,34 @@ describe('removeNPMAbsolutePaths.js', function () {
           expect(writeFile).to.have.been.calledOnce.and.calledWith(filePath);
         });
     });
+
+    it('checks that value of file contents do not contain specific removed fields', function () {
```
Comment (juanjoDiaz) [inline test/removeNPMAbsolutePaths.js]: The name should say what you are validating not how you validate it.

This test should go into 
```
describe('removeNPMAbsolutePaths.js', function () {
  ...
  describe('is valid file path', function () {
    ....
    it('removes only the specified fields if the --fields option is used', function () {
      <TEST_HERE>
```

File: test/removeNPMAbsolutePaths.js [hunk 9]
```diff
@@ -406,5 +403,34 @@ describe('removeNPMAbsolutePaths.js', function () {
...
+      writeFile.restore();
+      writeFile = sinon.stub(fs, 'writeFile').callsFake((path, fileContents, cb) => {
```
Comment (juanjoDiaz) [inline test/removeNPMAbsolutePaths.js]: The tests should always be:
```
Configure
Test
Assert
```

To do checks in the arguments send to stub you can use `writeFile.getCall(0).args[1]` to get the argument (not sure if the index should be 1 or other).
So, don't change the `writeFile` stub and just add bellow the line `expect(writeFile).to.have.been.calledOnce.and.calledWith(filePath)`:
```
        const packageJson = JSON.parse(writeFile.getCall(0).args[1]);
        expect(packageJson).to.not.have.property('_inBundle');
        expect(packageJson).to.not.have.property('_where');
        expect(packageJson).to.have.property('_from');
```

**Review by juanjoDiaz — CHANGES_REQUESTED** [review]
Thanks for the good work.
Almost there!

Few more comments :)

File: bin/removeNPMAbsolutePaths [hunk 10]
```diff
@@ -16,11 +16,30 @@ const opts = {
   force: false,
 };
 
-args.slice(1).forEach((arg) => {
-  if (arg === '--force') {
-    opts.force = true;
+let argWarning = [];
```
Comment (juanjoDiaz) [inline bin/removeNPMAbsolutePaths]: Please rename to `ignoredOptions`

File: bin/removeNPMAbsolutePaths [hunk 11]
```diff
@@ -16,11 +16,30 @@ const opts = {
...
+      opts.force = true;
+      break;
+    case '--fields':
+      opts.fields = opts.fields || [];
+      while (args[i + 1] && args[i + 1].slice(0, 2) !== '--') {
+        opts.fields.push(args[i += 1]);
+      }
+      break;
+    default:
+      argWarning = ['Invalid argument.', 'Extra arguments are being ignored'];
```
Comment (juanjoDiaz) [inline bin/removeNPMAbsolutePaths]: this should do `ignoredOptions.push(arg);` so at the end of the loop we get the list of ignore options/arguments

File: bin/removeNPMAbsolutePaths [hunk 12]
```diff
@@ -16,11 +16,30 @@ const opts = {
...
+      break;
   }
-});
+}
+
+if (argWarning.length) {
+  console.warn.apply(null, argWarning);
```
Comment (juanjoDiaz) [inline bin/removeNPMAbsolutePaths]: We can print the list with a nice message as:
`console.warn(`The following options are unknown and will be ignored:\n${ignoredOptions.join('\n')}`)`

File: test/data/specific_fields/module/package.json (hunk 2 above)
Comment (juanjoDiaz) [reply test/data/specific_fields/module/package.json]: Please remove this file.

**Review by juanjoDiaz — CHANGES_REQUESTED** [review]
Just some final touches!

File: test/removeNPMAbsolutePaths.js [hunk 13]
```diff
@@ -406,5 +428,6 @@ describe('removeNPMAbsolutePaths.js', function () {
           expect(writeFile).to.have.been.calledOnce.and.calledWith(filePath);
         });
     });
+
```
Comment (juanjoDiaz) [inline test/removeNPMAbsolutePaths.js]: I think that this empty line is failing the build because of linting error. Please fix it :)

**Review by juanjoDiaz — APPROVED** [review]
Great job!!

I'll merge and release as soon as I have a minute :)

**Comment by juanjoDiaz** [comment]
Resolves #7
